File: app/Core/Sources/EventDrivenContextCore/CaptureItem.swift
import Foundation

public enum CaptureStatus: String, Codable, CaseIterable, Sendable {
    case recording
    case interrupted
    case localSaved = "local_saved"
    case queued
    case waitingNetwork = "waiting_network"
    case waitingAuth = "waiting_auth"
    case uploading
    case synced
    case processing
    case processingFailed = "processing_failed"
    case ready
    case failed

    public var localizationKey: String { "capture_status_\(rawValue)" }
    public var canRetry: Bool {
        self == .interrupted || self == .failed || self == .waitingNetwork || self == .waitingAuth
    }
}

public struct CaptureItem: Codable, Equatable, Identifiable, Sendable {
    public var id: String { captureID }
    public let captureID: String
    public let accountID: String
    public let endpoint: String
    public let projectID: String
    public let projectName: String
    public let filePath: String
    public let mediaType: String
    public var sizeBytes: Int64
    public var sha256: String
    public let occurredAt: String
    /// The exact JSON string submitted on every retry.
    public let metadataJSON: String
    public var status: CaptureStatus
    public var eventID: String?
    public var remoteFileID: String?
    public var attemptCount: Int
    public var nextAttemptAt: Date?
    public var errorCode: String?
    public let createdAt: Date
    public var updatedAt: Date

    public init(
        captureID: String,
        accountID: String,
        endpoint: String,
        projectID: String,
        projectName: String,
        filePath: String,
        mediaType: String = "audio/mp4",
        sizeBytes: Int64 = 0,
        sha256: String = "",
        occurredAt: String,
        metadataJSON: String,
        status: CaptureStatus = .recording,
        eventID: String? = nil,
        remoteFileID: String? = nil,
        attemptCount: Int = 0,
        nextAttemptAt: Date? = nil,
        errorCode: String? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.captureID = captureID
        self.accountID = accountID
        self.endpoint = endpoint
        self.projectID = projectID
        self.projectName = projectName
        self.filePath = filePath
        self.mediaType = mediaType
        self.sizeBytes = sizeBytes
        self.sha256 = sha256
        self.occurredAt = occurredAt
        self.metadataJSON = metadataJSON
        self.status = status
        self.eventID = eventID
        self.remoteFileID = remoteFileID
        self.attemptCount = attemptCount
        self.nextAttemptAt = nextAttemptAt
        self.errorCode = errorCode
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    public var uploadIdentity: UploadIdentity {
        UploadIdentity(
            captureID: captureID,
            accountID: accountID,
            endpoint: endpoint,
            projectID: projectID,
            filePath: filePath,
            mediaType: mediaType,
            occurredAt: occurredAt,
            metadataJSON: metadataJSON,
            sha256: sha256
        )
    }

    public var canRetry: Bool {
        guard [.interrupted, .localSaved, .queued, .waitingNetwork, .waitingAuth, .uploading, .failed].contains(status) else {
            return false
        }
        return !["local_file_changed", "local_file_unreadable", "confirmation_mismatch",
                  "invalid_input", "conflict", "too_large", "request_too_large"]
            .contains(errorCode)
    }
}

/// Immutable fields that must remain identical when a lost response is retried.
public struct UploadIdentity: Codable, Equatable, Sendable {
    public let captureID: String
    public let accountID: String
    public let endpoint: String
    public let projectID: String
    public let filePath: String
    public let mediaType: String
    public let occurredAt: String
    public let metadataJSON: String
    public let sha256: String
}

File: app/android/app/src/main/java/life/integ/context/CaptureItem.java
package life.integ.context;

import java.io.File;

final class CaptureItem {
    static final String READY = "ready";
    static final String UPLOADING = "uploading";
    static final String EVENT_PENDING = "event_pending";
    static final String SYNCED = "synced";
    static final String INTERRUPTED = "interrupted";
    static final String AUTH_REQUIRED = "auth_required";
    static final String FAILED = "failed";

    final String id;
    final String endpoint;
    final String accountId;
    final String projectId;
    final String localPath;
    final String filename;
    final String mediaType;
    final long sizeBytes;
    final String sha256;
    final long durationMs;
    final String occurredAt;
    final String fileId;
    final String state;
    final String lastError;
    final long createdAt;

    CaptureItem(String id, String endpoint, String accountId, String projectId,
                String localPath, String filename, String mediaType, long sizeBytes,
                String sha256, long durationMs, String occurredAt, String fileId,
                String state, String lastError, long createdAt) {
        this.id = id;
        this.endpoint = endpoint;
        this.accountId = accountId;
        this.projectId = projectId;
        this.localPath = localPath;
        this.filename = filename;
        this.mediaType = mediaType;
        this.sizeBytes = sizeBytes;
        this.sha256 = sha256;
        this.durationMs = durationMs;
        this.occurredAt = occurredAt;
        this.fileId = fileId;
        this.state = state;
        this.lastError = lastError;
        this.createdAt = createdAt;
    }

    boolean fileStillMatches() {
        File file = new File(localPath);
        return file.isFile() && file.length() == sizeBytes;
    }

    boolean belongsToAccount(SessionStore.Session session) {
        return session != null
                && endpoint.equals(session.endpoint)
                && accountId.equals(session.userId);
    }
}
